pepfile: "project_config.yaml"
configfile: "config.yaml"

kSize=config["kSize"]
outFolder=config["outFolder"]
tmpFolder=config["tmpFolder"]

references=list(filter(lambda x:x.library=="parent_reference",pep.samples))
references=[(x.sample_name,x.file[0]) for x in references]
references=dict(references)


progny_short=list(filter(lambda x:x.library=="progny_short",pep.samples))
progny_short=[(x.sample_name,x.file) for x in progny_short]
progny_short=dict(progny_short)
dbgs=list(references.keys())+list(progny_short.keys())

sampleParents={}

for s in list(filter(lambda x:x.library=="progny_short",pep.samples)):
    parents=s.parent.split("x")
    sampleParents[s.sample_name]=[]
    for p in parents:
    	sampleParents[s.sample_name].append(outFolder+"corticall/"+(s.sample_name+"."+p)+".ref.ctp.gz")


def getFile(dict,name):
    return dict[name]

import os,math

def getFQSizeInGB(name):
    files=progny_short[name]
    res=0
    for f in files:
    	res+=os.path.getsize(f)
    return int(math.ceil(float(res)/(1024.0*1024.0*1024.0)))

def getFileSizeInGB(file):
    res = os.path.getsize(file)
    return int(math.ceil(float(res)/(1024.0*1024.0*1024.0)))

rule all:
    input:
        expand(outFolder+"corticall/{sample}.ctx",sample=dbgs),
        expand(outFolder+"corticall/{sample}.cleaned.edgesInfered.popped.ctx",sample=progny_short.keys()),
        expand(outFolder+"kmc/{sample}.kmc_pre",sample=dbgs),
     #   expand(outFolder+"corticall/{sampleParent}.ref.ctp.gz", sample=sampleParent), 
     #   expand(outFolder+"corticall/{sample}.se.ctp.gz",sample=progny_short.keys()),
#        expand(outFolder+"corticall/{sample}.contigs.fa",sample=progny_short.keys()),
        expand(outFolder+"bcalm/{sample}.unitigs.fa",sample=progny_short.keys()),
        expand(outFolder+"kProcessor/{sample}.kframe.cleaned.phmap",sample=progny_short.keys())


rule buildParentDBG:
    input:  lambda wildcards: getFile(references,f"{wildcards.sample}")
    output:
        ctx=outFolder+"corticall/{sample}.ctx",
        time=outFolder+"corticall/{sample}.build.time"
    log: outFolder+"corticall/{sample}.build.log"
    conda: "env.yaml"
    shell:
       	"""
	       /usr/bin/time -v -o {output.time} mccortex {kSize} build -k {kSize} -t 1 --sample {wildcards.sample} -S -m 4G -n 100M  -1 {input} {output.ctx} &> {log}
        """

rule buildPrognyDBG:
    input:  lambda wildcards: getFile(progny_short,f"{wildcards.sample}")
    output:
        ctx=outFolder+"corticall/{sample}.ctx",
        time=outFolder+"corticall/{sample}.build.time"
    log: outFolder+"corticall/{sample}.build.log"
    conda: "env.yaml"
    resources: mem_gb= lambda wildcards: getFQSizeInGB(f"{wildcards.sample}") * 8
    shell:
       	"""
	       /usr/bin/time -v -o {output.time} mccortex {kSize} build -k {kSize} -t 3 --sample {wildcards.sample} -S -m {resources.mem_gb}G  -2 {input[0]}:{input[1]} {output.ctx} &> {log}
        """

rule cleanPrognyDBG:
    input:  outFolder+"corticall/{sample}.ctx"
    output:
        ctx=outFolder+"corticall/{sample}.cleaned.ctx",
        covg_after=outFolder+"corticall/{sample}.cleaned.covg_after",
        covg_before=outFolder+"corticall/{sample}.cleaned.covg_before",
        len_after=outFolder+"corticall/{sample}.cleaned.len_after",
        len_before=outFolder+"corticall/{sample}.cleaned.len_before",
        time=outFolder+"corticall/{sample}.clean.time"
    log: outFolder+"corticall/{sample}.clean.log"
    conda: "env.yaml"
    resources: mem_gb= lambda wildcards: getFileSizeInGB(outFolder+f"corticall/{wildcards.sample}.ctx") * 4
    shell:
       	"""
	       /usr/bin/time -v -o {output.time} mccortex {kSize} clean -t 1  -m {resources.mem_gb}G -o {output.ctx} -S  -c {output.covg_before} -C {output.covg_after} -l {output.len_before} -L {output.len_after} {input} &> {log}
        """


rule inferEdgesPrognyDBG:
    input:  outFolder+"corticall/{sample}.cleaned.ctx"
    output:
        ctx=outFolder+"corticall/{sample}.cleaned.edgesInfered.ctx",
        time=outFolder+"corticall/{sample}.inferedges.time"
    log: outFolder+"corticall/{sample}.inferedges.log"
    conda: "env.yaml"
    shell:
       	"""
	       /usr/bin/time -v -o {output.time} mccortex {kSize} inferedges -m 4G -o {output.ctx} {input} &> {log}
        """


rule popBubblesPrognyDBG:
    input:  outFolder+"corticall/{sample}.cleaned.edgesInfered.ctx"
    output:
        ctx=outFolder+"corticall/{sample}.cleaned.edgesInfered.popped.ctx",
        time=outFolder+"corticall/{sample}.popBubbles.time"
    log: outFolder+"corticall/{sample}.popBubbles.log"
    conda: "env.yaml"
    resources: mem_gb= lambda wildcards: getFileSizeInGB(outFolder+f"corticall/{wildcards.sample}.cleaned.edgesInfered.ctx") * 2
    shell:
       	"""
	       /usr/bin/time -v -o {output.time} mccortex {kSize} popbubbles  -m {resources.mem_gb}G -o {output.ctx}  {input} &> {log}
        """


rule threadSEProgny:
    input:
        ctx=outFolder+"corticall/{sample}.cleaned.edgesInfered.ctx",
        reads=lambda wildcards: getFile(progny_short,f"{wildcards.sample}")
    output:
        ctp=outFolder+"corticall/{sample}.se.ctp.gz",
        time=outFolder+"corticall/{sample}.threadSE.time"
    log: outFolder+"corticall/{sample}.threadSE.log"
    conda: "env.yaml"
    shell:
       	"""
       	# arguments meaining:
       	# -W Use two-way gap filling (liberal)
       	# -E Skip extra check after gap bridging
	       /usr/bin/time -v -o {output.time} mccortex {kSize} thread  -t1 -W -E -m 4G -o {output.ctp} -1 {input.reads[0]} -1 {input.reads[1]} {input.ctx} &> {log}
        """

rule threadParentRef:
    input:
        ctx=outFolder+"corticall/{sample}.cleaned.edgesInfered.ctx",
        ref= lambda wildcards: getFile(references,f"{wildcards.ref}")
    output:
        ctp=outFolder+"corticall/{sample}.{ref}.ref.ctp.gz",
        time=outFolder+"corticall/{sample}.{ref}.threadParent.time"
    log: outFolder+"corticall/{sample}.{ref}.threadParent.log"
    conda: "env.yaml"
    shell:
       	"""
       	# arguments meaining:
       	# -W Use two-way gap filling (liberal)
       	# -E Skip extra check after gap bridging
	       /usr/bin/time -v -o {output.time} mccortex {kSize} thread  -t1 -W -E -m 4G -o {output.ctp} -1 {input.ref}  {input.ctx} &> {log}
        """


rule contigs:
    input:
        ctx=outFolder+"corticall/{sample}.cleaned.edgesInfered.ctx",
        ctpSE=outFolder+"corticall/{sample}.se.ctp.gz",
        ctpRef= lambda wildcards: getFile(sampleParents,f"{wildcards.sample}")
    output:
        contigs=outFolder+"corticall/{sample}.contigs.fa",
        time=outFolder+"corticall/{sample}.contigs.time"
    log: outFolder+"corticall/{sample}.contigs.log"
    conda: "env.yaml"
    shell:
       	"""
       	# arguments meaining:
       	# -M  Do not use the missing information check
       	# -G Genome size in bases

	    /usr/bin/time -v -o {output.time} mccortex {kSize} contigs  -t1 -m 4G -G 23332839 -M -o {output.contigs} -p {input.ctpRef[0]} -p {input.ctpRef[1]} -p {input.ctpSE} {input.ctx} &> {log}
        """



rule kmcParent:
    input:  lambda wildcards: getFile(references,f"{wildcards.sample}")
    output:
         pre=outFolder+"kmc/{sample}.kmc_pre",
         suf=outFolder+"kmc/{sample}.kmc_suf",
         time=outFolder+"kmc/{sample}.kmc.time"
    threads: 4
    log: outFolder+"kmc/{sample}.kmc.log"
    shell:
       	"""
	mkdir -p {tmpFolder}$$/
	/usr/bin/time -v -o {output.time} KMC/bin/kmc -ci1 -t{threads} -k{kSize} -m4 -fm {input} {outFolder}kmc/{wildcards.sample}  {tmpFolder}$$/ &> {log}
        rm -rf {tmpFolder}$$/
	"""


rule kmcProgny:
    input: lambda wildcards: getFile(progny_short,f"{wildcards.experiment}")
    output:
         pre=outFolder+"kmc/{experiment}.kmc_pre",
         suf=outFolder+"kmc/{experiment}.kmc_suf",
         time=outFolder+"kmc/{experiment}.kmc.time"
    conda:
         "env.yaml"
    threads: 4
    log: outFolder+"kmc/{experiment}.kmc.log"
    shell:
       	"""
	    mkdir -p {tmpFolder}$$/
    	    echo {input} |tr -s ' ' $'\n' > {wildcards.experiment}.lst
	        /usr/bin/time -v -o {output.time} KMC/bin/kmc -ci1 -cs1000000  -hp  -t{threads} -k{kSize} -m4  @{wildcards.experiment}.lst {outFolder}kmc/{wildcards.experiment}  {tmpFolder}$$/ &> {log}
		rm -rf {tmpFolder}$$/
	     #   rm {wildcards.experiment}.lst
        """

rule buildPrognyDBGKprocessor:
    input:  kmc= outFolder+"kmc/{sample}.kmc_pre"
    output:
        ctx=outFolder+"kProcessor/{sample}.kframe.phmap",
        time=outFolder+"kProcessor/{sample}.kprocessor.build.time"
    log: outFolder+"kProcessor/{sample}.kprocessor.build.log"
    conda: "env.yaml"
    shell:
       	"""
	       /usr/bin/time -v -o {output.time} ../../apps/loadFromKMC {outFolder}kmc/{wildcards.sample} PHMAP {outFolder}kProcessor/{wildcards.sample}.kframe  &> {log}
        """


rule bcalmProgny:
    input: lambda wildcards: getFile(progny_short,f"{wildcards.experiment}")
    output:
         unitigs=outFolder+"bcalm/{experiment}.unitigs.fa",
         time=outFolder+"bcalm/{experiment}.bcalm.time"
    conda:
         "env.yaml"
    threads: 4
    log: outFolder+"bcalm/{experiment}.bcalm.log"
    shell:
       	"""
    	    echo {input} |tr -s ' ' $'\n' > {wildcards.experiment}.lst
    	    /usr/bin/time -v -o {output.time} bcalm -nb-cores {threads} -in {wildcards.experiment}.lst -kmer-size {kSize} -abundance-min 1 -out {outFolder}bcalm/{wildcards.experiment} &> {log}
        """


rule cleanPrognyDBGKprocessor:
    input:
        kframe=outFolder+"kProcessor/{sample}.kframe.phmap",
        unitigs=outFolder+"bcalm/{sample}.unitigs.fa"
    params:
        kframeINPrefix=outFolder+"kProcessor/{sample}.kframe",
        kframeOUTPrefix=outFolder+"kProcessor/{sample}.kframe.cleaned"
    output:
        kframe=outFolder+"kProcessor/{sample}.kframe.cleaned.phmap",
        time=outFolder+"kProcessor/{sample}.kprocessor.clean.time"
    log: outFolder+"kProcessor/{sample}.kprocessor.clean.log"
    conda: "env.yaml"
    shell:
       	"""
	       /usr/bin/time -v -o {output.time} ./cleanDBG -k  {params.kframeINPrefix} -u {input.unitigs} -o {params.kframeOUTPrefix}  &> {log}
        """

